### DISCUSSION TITLE
Implemented Solvers for Finite Horizon POMDP Problem

### DISCUSSION THREAD

**Discussion by turguy** [JuliaPOMDP/POMDPs.jl#460] [Problem Implementation]
Hello everyone,

I am conducting a research project and have developed a 7-tuple for my POMDP problem, and implemented the problem using the Julia POMDP interface (QuickPOMDP). Using solvers such as SARSOP or Point Based Value Iteration, I was able to obtain a policy for the infinite horizon POMDP.

However, I now need to convert this problem into a finite horizon POMDP and obtain a policy for this as well. To achieve this, I am using the "FiniteHorizonPOMDPs.jl" package and testing with a horizon length of 2. With just one line of code, I can change the infinite horizon problem to a finite horizon problem (finite_POMDP = fixhorizon(infinite_POMDP, 2)).
But I am encountering difficulties with obtaining a policy using this method. When using SARSOP, I received an error message: ERROR
  model.pomdpx:Line 129076:
  In Observation Function observation : there are missing transitions
  Transition not specified for instance a0 s0 
.

 And when using the Point Based Value Iteration solver, the model did not load even after running it for a day. This is surprising, as with the same setup for the infinite horizon, I was able to obtain policies using these solvers in less than an hour (for testing purposes, I minimized the size of the problem).
I would greatly appreciate any suggestions or ideas on what I may be doing wrong.

**Comment by qhho** [comment]
I believe both the implementations of the Julia wrapper for SARSOP in C++ and Point Based Value Iteration do not currently work well with finite horizon POMDPs. For finite horizon MDPs, there is https://github.com/JuliaPOMDP/FiniteHorizonValueIteration.jl

Edit: Could you try using NativeSARSOP (https://github.com/JuliaPOMDP/NativeSARSOP.jl) and see if the issue persists?

@WhiffleFish perhaps we can add functionality for the finite horizon case for NativeSARSOP

**Reply by zsunberg** [reply]
Sorry if this was not clear. *Any* POMDP solver *should* work with a finite horizon POMDP. `fixhorizon` augments the state with time. This is just a matter of tracking down the bug.

**Reply by zsunberg** [reply]
(you may be able to make NativeSARSOP have a more optimized version for finite horizons though)

**Reply by qhho** [reply]
Sorry, I was mistaken, Point Based Value Iteration and NativeSARSOP should work in general with finite horizon POMDPs. There is an issue with SARSOP.jl regarding the model.pomdpx file that seems to happen for all finite horizon POMDPs. This could be a bug.

**Reply by qhho** [reply]
@turguy Could you share more about the specific POMDP problem you have implemented?

**Reply by turguy** [reply]
Hi, thank you for all the replies. Here is [the notebook](https://colostate-my.sharepoint.com/:u:/g/personal/tcaglar_colostate_edu/EVJatBgXolxOpoHYYJvGCK0BG4Ub3YF6j_Zzi8EjZPBh5A?e=7VIluL) that I used for my POMDP problem

**Reply by WhiffleFish** [reply]
Hey @turguy, FiniteHorizonPOMDPs.jl has been fixed, so you should be able to solve your finite horizon POMDP now.

**Reply by turguy** [reply]
Thanks for letting me know! I appreciate the update. You're the best!
